cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/SourceControlContext.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.sourcecontrol;
+
+import io.cdap.cdap.api.security.store.SecureStore;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryConfig;
+
+import java.io.File;
+import java.nio.file.Path;
+
+public class SourceControlContext {
+  // Default value for git timeout.
+  private static final int GIT_COMMAND_TIMEOUT_SECONDS_DEFAULT = 30;
+  // Default value for git repo
+  private static final String GIT_REPOSITORIES_CLONE_DIRECTORY_PATH_DEFAULT = "/tmp/source-control/";
+  private final SecureStore store;
+  private final Path repoBasePath;
+  private final Path validationRepoPath;
+  private final RepositoryConfig repositoryConfig;
+
+  private final int gitCommandTimeoutSeconds;
+  private final NamespaceId namespaceId;
+
+  public SourceControlContext(SecureStore store,
+                              RepositoryConfig repositoryConfig,
+                              NamespaceId namespaceId,
+                              CConfiguration cConf) {
+    this.store = store;
+    this.repositoryConfig = repositoryConfig;
+    this.namespaceId = namespaceId;
+    this.gitCommandTimeoutSeconds =
+      cConf.getInt(Constants.SourceControl.GIT_COMMAND_TIMEOUT_SECONDS, GIT_COMMAND_TIMEOUT_SECONDS_DEFAULT);
+    String gitCloneDirectory = cConf.get(Constants.SourceControl.GIT_REPOSITORIES_CLONE_DIRECTORY_PATH,
+                                         GIT_REPOSITORIES_CLONE_DIRECTORY_PATH_DEFAULT);
+    Path commonPath = new File(gitCloneDirectory).toPath().resolve("namespace").resolve(namespaceId.getNamespace());
+    this.repoBasePath = commonPath.resolve("actual");
+    this.validationRepoPath = commonPath.resolve("validate");
```

Comment:
If the SCM already exists and is initialized the main directory would have data. Calling validate will clean up the directory. So we have 3 options:
1. Disallow validate on initialized repos.
2. Use a different directory for validation and cloning.
3. Use the initialized git client for doing an LsRemote. Doing this would require ensuring validation doesn't change the state of the existing git client.
So I chose to have separate directories for validation. Now I've moved the responsibility of creating 2 sub-directories into SCM so SCContext provides only one directory.